compiler/rustc_mir_transform/src/shim.rs

```diff
@@ -1005,6 +1013,609 @@ fn build_fn_ptr_addr_shim<'tcx>(tcx: TyCtxt<'tcx>, def_id: DefId, self_ty: Ty<'t
+        let (size, _) = discr.ty.int_size_and_signed(self.tcx);
+        self.put_operand(Operand::const_from_scalar(
+            self.tcx,
+            discr.ty,
+            interpret::Scalar::from_uint(discr.val, size),
+            self.span,
+        ));
+    }
+
+    /// Puts `x: RvalueType` on top of the stack.
+    fn put_temp_rvalue(&mut self, rvalue: Rvalue<'tcx>) {
+        let last_bb = &mut self.bbs[self.last_bb];
+        debug_assert!(last_bb.terminator.is_none());
+        let source_info = self.source_info;
+
+        let local_ty = rvalue.ty(&self.locals, self.tcx);
+        // We need to create a new local to be able to "consume" it with
+        // a combinator
+        let local = self.locals.push(LocalDecl::with_source_info(local_ty, source_info));
+        last_bb.statements.extend_from_slice(&[
+            Statement { source_info, kind: StatementKind::StorageLive(local) },
+            Statement {
+                source_info,
+                kind: StatementKind::Assign(Box::new((local.into(), rvalue))),
+            },
+        ]);
+
+        self.put_operand(Operand::Move(local.into()));
+    }
+
+    /// Puts operand on top of the stack.
+    fn put_operand(&mut self, operand: Operand<'tcx>) {
+        if let Some(top_cleanup_bb) = &mut self.top_cleanup_bb {
+            let source_info = self.source_info;
+            self.bbs.raw.reserve(2);
+            match &operand {
+                Operand::Copy(_) | Operand::Constant(_) => {
+                    for _ in 0..2 {
+                        *top_cleanup_bb = self.bbs.push(BasicBlockData {
+                            statements: Vec::new(),
+                            terminator: Some(Terminator {
+                                source_info,
+                                kind: TerminatorKind::Goto { target: *top_cleanup_bb },
+                            }),
+                            is_cleanup: true,
+                        });
+                    }
+                }
+                Operand::Move(place) => {
+                    let local = place.as_local().unwrap();
+                    *top_cleanup_bb = self.bbs.push(BasicBlockData {
+                        statements: vec![Statement {
+                            source_info,
+                            kind: StatementKind::StorageDead(local),
+                        }],
+                        terminator: Some(Terminator {
+                            source_info,
+                            kind: TerminatorKind::Goto { target: *top_cleanup_bb },
+                        }),
+                        is_cleanup: true,
+                    });
+                    *top_cleanup_bb = self.bbs.push(BasicBlockData {
+                        statements: Vec::new(),
+                        terminator: Some(Terminator {
+                            source_info,
+                            kind: if self.locals[local].ty.needs_drop(self.tcx, self.param_env) {
```

Comment:
leave a FIXME so we look at this again to see if we should insert a runtime check of `needs_drop` here so that LLVM can figure out how to DCE it after monomorphization (in case `ty` is generic)

Reply:
I thought shims are constructed after monomorphization (in other words without generics), but ok.

Reply:
Should I revert it? a14ae58959187bbcc91f7b6db9061b6e857a6a90